[Fallman2] iP

Fallman2 opened the pull request:
# **Puke**
Have you ever been reading something and randomly thought "You know what'd be funny? If this text were all written like that 'Expanding Brain' meme.". No? Well too bad because that's what this is.

```
Salutations! I hereby would like to inform you that my identity is that of
 ____        _        
|  _ \ _   _| | _____ 
| |_| | | | | |/ / _ \
| ___/| |_| |    | __/
| |    \__,_|_|\_\___|
|_|
An exceedingly verbose conversation simulation program.
```
Puke allows you to spend less time remembering what you're supposed to be doing, allowing you to spend your saved time on what truly matters: Melting your brain with needlessly complicated text.

Features:
- Text based, because kids these days don't do nearly enough reading
- So fast that it won't ask for confirmation, allowing you to easily mistakenly delete your entire to do list
- Copious amount of words because that definitely helps any given situation

```
Understood. I have hereby created a task known to require doing at a future time but with no such time being specified and inserted it into the overall collection of said tasks that require action.
Here is a display of the added deadline task: [T][ ]  borrow book
You now, in total, have 1 of these tasks recorded within said collection.
```
General Functions:

- [x] Store a list of your tasks that can be easily accessed
- [x] Search for tasks that match keywords provided
- [x] Remind yourself how naggy your parents can be

Steps to download

0. Make sure you have Java installed
1. Download it [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/406)
2. Run it

Lastly, if you've made it this far, here's a quote for your troubles:
> "Outside of a dog, a book is a man's best friend. Inside of a dog, it's too dark to read." - Groucho Marx

And if you're a programmer, you probably don't wanna examine the code too closely, lest you form an aneurysm. There isn't even a GUI yet. Here's what the `main` method looks like:
```
public static void main(String[] args) throws IOException {
        new Puke().run();
    }
```
Don't say I didn't warn you. :man_shrugging:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Fliprun committed:
Level-0

Fliprun committed:
Level-1

Fliprun committed:
Level-2

Fliprun committed:
Level-3

Fliprun committed:
Level-4

Fliprun committed:
A-TextUiTesting

Fliprun committed:
Level-5

Fliprun committed:
Level-6

Fliprun committed:
Touching Up

Fliprun committed:
Update file name and edited isEmpty() condition check

File name was previously Duke, unchanged from the original despite the change in the name since the first commit. There was also a condition check that could have been shortened.

File name has been appropriately changed and the condition has been updated.

Fliprun committed:
Create new class to handle data

The chatbot previously did not have a way to handle storing data after closing.

To prepare for the implementationn of data storage, a text file has been created alongside a directory and a new class, DataHandler is created to translate this file into
respective Tasks.

Fliprun committed:
Complete saving function

The saving function was previously only implemented partiallly. There were also some display bugs where the separator line was not printed where it should have been.

Completed the saving function and fixed the bugs.

Fliprun committed:
Merge branch 'branch-Level-7'

Fliprun committed:
Implement LocalDateTime Support

Events and deadlines did not properly use dates and times in their implementation previously, allowing for unmeaningful dates and times.

LocalDateTime has been implemented and some small String irregularities have been fixed

Fliprun committed:
Merge branch 'branch-Level-8'

Fliprun committed:
Add new classes for extra OOP

Previous implementation had too many steps crammed into the main method, leading to messy code.

Added Command classes and subclasses to categorize the commands.

Fliprun committed:
Add Ui class and fix errors in command classes

Ui displays are still executed in main class and could be furthur separated with oop

Created the UI class to store all UI messages.

Fliprun committed:
Update main method in Puke class

After all the classes were changed in the previous commit, the main class no longer ran properly.

Changed the main class too run properly with the new coode

Fliprun committed:
Add files to a package

Fliprun committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Fliprun committed:
Set up Gradle as per instructions provided

Fliprun committed:
Add sample JUnit tests

Fliprun committed:
Edit build.gradle to properly save a jar file

Fliprun committed:
Update Puke.java to fit coding standard

Fliprun committed:
Add new find feature

There was no way to look for specific tasks previously

Added a way to filter tasks

Fliprun committed:
Add comments for proper Javadoc documentation

Fliprun committed:
Merge branch 'branch-A-CodingStandard'

Fliprun committed:
Merge branch 'branch-Level-9'

Fliprun committed:
Merge branch 'branch-A-JavaDoc'

#	src/main/java/puke/Puke.java

Fliprun committed:
Correct grammar errors in Javadoc comments

Certain grammatically incorrect words were being used previously in the Javadoc comments

Changed these words to be grammatically correct

Fliprun committed:
Implement Checkstyle

Previously, it was difficult to check through the full length of code for code standard violations

Checkstyle is implemented to automatically detect these violations

Fliprun committed:
Merge branch 'branch-A-CheckStyle'

Fliprun committed:
Create and sort each class into its proper package

Previously, all classes were in the same file and thus in the same package, leading to messier coding.

Created new folders and packages and sorted each class into its appropriate package.

J-hta-n commented on src/main/java/puke/DataHandler.java:
```diff
@@ -0,0 +1,102 @@
+import puke.task.Deadline;
+import puke.task.Event;
+import puke.task.Task;
+import puke.task.ToDo;
```
I like how the ordering of import statements are consistent and organised 👍

J-hta-n commented on src/main/java/puke/DataHandler.java:
```diff
@@ -0,0 +1,102 @@
+     * @throws PukeException If an invalid task is detected.
+     */
+    public static Task translate(String input) throws PukeException {
+        String[] split = input.split("/");
```
Perhaps can have a variable name in plural form eg inputParts? 💅

J-hta-n commented on src/main/java/puke/Puke.java:
```diff
@@ -0,0 +1,74 @@
+
+
+/**
+ * A chatbot that uses overly complicated sentences.
```
I like how clean the Puke.java code is 🍁

J-hta-n commented on src/main/java/puke/TaskList.java:
```diff
@@ -0,0 +1,118 @@
+import puke.task.Task;
+
+/**
+ * A class that stores Tasks in a list.
```
Naming, comments and layouts adhere well to coding standards in general 🌲

J-hta-n commented on src/test/java/puke/PukeTest.java:
```diff
@@ -0,0 +1,42 @@
+        //Testing Parsing DeadlineCommand
+        assertEquals(new DeadlineCommand("return book /by 2019-12-01T10:00"), Parser.parse("deadline", " return book /by 2019-12-01T10:00"));
+        //Testing Parsing EventCommand
+        assertEquals(new EventCommand("project meeting /from 2023-08-30T14:00 /to 2023-08-30T16:00"), Parser.parse("event", " project meeting /from 2023-08-30T14:00 /to 2023-08-30T16:00"));
```
Should this be line be wrapped? 🍇

J-hta-n commented on src/main/java/puke/Ui.java:
```diff
@@ -0,0 +1,221 @@
+import puke.task.Task;
+
+/**
+ * A class that handles all messages printed in the UI
```
Perhaps can rename some of the method names as verbs instead of nouns (eg getMsg() instead of msg())? 🥝

J-hta-n reviewed (commented):
- Overall well documented, PR description was ***beautiful***, and code base is neat :)
- Just a few minor pointers on naming and layout 👍

Fliprun committed:
Create new sub-package

Previously, there were still many files not in a sub package, leading to messy file organization

Created a new package "managers" in order to store classes that handle data. Also created a dummy method in preparation for GUI creation.

Fliprun committed:
Add in barebones GUI

Previously, there was no GUI functionality

Added in a barebones GUI to be improved in the next commit

Fliprun committed:
Implement functioning GUI

Previously the GUI implementation was barebones and the UI class was handling messages displayed in the text based UI.

Changed the command classes to generate the messages required by the UI, rendering the UI class obselete. UI class has been deleted and a more functional GUI was implemented that managed to output the messages previously used by Puke.

Fliprun committed:
Merge branch 'branch-Level-10'

Fliprun committed:
Readd startup message

The startup message had been removed previously while implementing the GUI

Startup message has been re-added

AryanG01 commented on src/main/java/puke/Puke.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * List of tasks stored by the chatbot
+     */
+    private TaskList tasks;
```
I like how you abstracted the list of tasks instead of using an ArrayList directly

AryanG01 commented on src/main/java/puke/Puke.java:
```diff
@@ -0,0 +1,77 @@
+            tasks = new TaskList(DataHandler.loadDatabase());
+        } catch (PukeException e) {
+            new File("ListData.txt").createNewFile();
+            tasks = new TaskList();
```
Nice error handling to create a new file if the code cant find one

AryanG01 commented on src/main/java/puke/Puke.java:
```diff
@@ -0,0 +1,77 @@
+    public static void main(String[] args) throws IOException {
+        new Puke().run();
+    }
+*/
```
Is this supposed to be a typo?

AryanG01 commented on src/main/java/puke/command/ClearCommand.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * A Command class that when executed, clears the task list.
+ */
+public class ClearCommand extends Command {
```
Nice additional feature added

AryanG01 commented on src/main/java/puke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,78 @@
+     * @return A boolean.
+     */
+    @Override
+    public boolean equals(Object other) {
```
A suggestion here would be to change the name to follow the naming conventions for a boolean name

AryanG01 commented on src/main/java/puke/command/MarkCommand.java:
```diff
@@ -0,0 +1,63 @@
+     * @param tl The task list.
+     * @return the message String.
+     */
+    public String execute(TaskList tl) {
```
shouldn't there be a throws keyword here?

AryanG01 commented on src/main/java/puke/command/TodoCommand.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * A Command class that when executed, creates a new Todo task
+ */
+public class TodoCommand extends Command {
```
Just a suggestion but would you like to add a date functionality here?

AryanG01 commented on src/main/java/puke/command/TodoCommand.java:
```diff
@@ -0,0 +1,77 @@
+package puke.command;
```
Instead of putting Todo, Deadline and Event under command do you think it would be better to put them under another category called tasks

AryanG01 commented on src/main/java/puke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,65 @@
+     * @param tl The task list.
+     * @return the message String.
+     */
+    public String execute(TaskList tl) {
```
You might want to use a StringBuilder here as it is faster than adding strings

AryanG01 commented on src/main/java/puke/managers/Parser.java:
```diff
@@ -0,0 +1,57 @@
+     * @return Corresponding command
+     * @throws PukeException If an invalid command or line is parsed
+     */
+    public static Command parse(String command, String line) throws PukeException {
```
A suggestion would be if a switch statement might make it easier for you

Fliprun committed:
Add assert statements to document assumptions

There are no assert statements which causes some incorrect inputs
to not trigger an exception. There is also an unused UI class
that was previously used to output responses.

Assert statements are added to document assumptions and prevent
more incorrect inputs which cause exceptions. The unused UI class
was also deleted.

Fliprun committed:
Add more assert statements to document more assumptions

Some classes had assumptions but did not have assert statements to
document them.

More assert statements are added in these classes to document the
assumptions.

Fliprun committed:
Refactored segments of code to improve code quality

There were many message strings that could have been static
variables.

Changed these Strings to static variables in their respective
classes.

Fallman2 committed:
Merge pull request #2 from Fallman2/Branch-A-CodeQuality

A-CodeQuality

Fallman2 committed:
Merge pull request #1 from Fallman2/branch-A-Assertions

Branch-A-assertions

Fliprun committed:
Remove unused segment of code

There is a segment of code that was previously used to run the
program before the GUI was implemented.

Removed this segment of code to improve readability.

Fliprun committed:
Add tagging function

There was no tagging function for the tasks.

Added tagging function.

Fliprun committed:
Change variable names to fit coding style

Variable names for some Booleans did not follow coding style.

Changed these names to better describe their function.

Some of these Booleans were also left unused after changing the UI
to a GUI. This should be addressed when the GUI is improved.

Fliprun committed:
Improve GUI formatting

The GUI was previously not very pleasant to look at and did not have
any indicators when an error occurred.

Added extra dialog box formatting to change colour depending on
whether an error occurs and ensured that the full textbox appears
when longer messages are sent.

Fliprun committed:
Merge branch 'branch-A-BetterGUI'

Fliprun committed:
Add more JUnit tests

Many JUnit tests implemented previously were rendered obsolete due
to some classes being deleted,

Added more JUnit tests to make up for it.

Fliprun committed:
Merge branch 'branch-A-MoreTesting'

Fliprun committed:
Add bot name in GUI window

GUI window was previously empty.

Added the name of the bot in the window and added a screenshot of
the bot in preparation for creation of a User Guide.

Fliprun committed:
Edit User Guide file

User Guide file was not properly set up.

Edited the User Guide file in preparation for finalizing the bot.

Fliprun committed:
Fix formatting on User Guide

User guide previously had unnecessary components which diminished
the formatting.

Removed certain sections to improve formatting.

Fliprun committed:
Fix exit command in GUI

The exit command previously did not work after implementing the GUI.

Fixed the command and allowed it to close the program when executed.
Also removed the launcher class and changed the GUI to launch from
the puke class instead.
